### DISCUSSION TITLE
Feature Request: Permit index creation on arrow-backed tables

### DISCUSSION THREAD

**Discussion by spenczar** [duckdb/duckdb#8300] [Ideas]
I'm working with moderately sized (~millions of rows) Arrow arrays that have a relational structure. I have a big table of `observations`, which has a Arrow string-typed column of `observer_id`s. I also have a big table of `observers` with (you guessed it) IDs that match the first table.

These tables are very wide, with hundreds of columns. My particular problem involves pulling out a small subset of those hundreds of columns from each table, joined on the observer ID. I have to do that very often in my application for selective subsets based on some features in the "observations" table. 

So this sounds a lot like I'd benefit from indexing the observer_id columns in my tables.

My tables are currently being constructed using `duckdb.arrow(...)`. Unfortunately, that means I cannot create indexes for those tables. In Python, I get this error:
```
BinderException: Binder Error: Can only create an index over a base table!
```

The arrow-backed tables have a type of "SUBQUERY_RELATION". 

A workaround is to copy all the data - something like
```
duckdb.sql("create table observations_base as select * from observations")
duckdb.sql("create index observations_idx on observations_base (obs_id)")
```
But this copy is quite unfortunate - it costs a lot of memory, and I was hoping for a zero-copy Arrow interface. It would be much better to be able to create an index directly on the Arrow data.

**Comment by Tishj** [comment]
I believe `duckdb.arrow()` creates a relation from an arrow object, by creating a stream, so we can lazily scan the arrow data.
```c++
	auto rel = connection
	               ->TableFunction("arrow_scan", {Value::POINTER(CastPointerToValue(stream_factory.get())),
	                                              Value::POINTER(CastPointerToValue(stream_factory_produce)),
	                                              Value::POINTER(CastPointerToValue(stream_factory_get_schema))})
	               ->Alias(name);
```

I'm not really clear on what you mean by
> My tables are currently being constructed using duckdb.arrow(...)

And I'm not sure what code you're running to result in
> BinderException: Binder Error: Can only create an index over a base table!

`duckdb.arrow()` creates a relation, which is not the same as a table.
It's essentially a scan over an arrow stream, which is why it's a `SUBQUERY_RELATION`.

You could create tables from this, explicitly through `duckdb.sql('create table tbl as select * from <name_of_your_relation>')`, at which point I believe you can add constraints on them (though I'm a little fuzzy on whether we added `ALTER TABLE ADD CONSTRAINT` support yet)

(I'm not 100% sure of this last part, so take it with a grain of salt)
I think creating tables from the scan does come at a cost though, because while arrow->duckdb vectors are mostly zero-copy, tables are stored in our storage format, which can't benefit from this zero-copy behavior

**Reply by spenczar** [reply]
Thanks for the corrections, @Tishj. Yes, my language was sloppy; I am creating relations and not tables.

I am not trying to add a constraint. I am trying to create an index for that relation.

Something like this is what I want to do, and what raises the cited error:

    >>> rel = duckdb.arrow(some_arrow_table)
    >>> duckdb.sql("create index on rel(id_column)")

**Reply by Tishj** [reply]
Ah right, you asked about Indexes, not constraints, I mixed them up because they are pretty similar.

I wonder if you could make a view out of the relation and create an Index on that, but if that doesn't work then I think the only option is to create a table from the arrow data and create the index on top of that
